gtsam/hybrid/HybridGaussianConditional.cpp

```diff
@@ -288,28 +310,39 @@ HybridGaussianConditional::shared_ptr HybridGaussianConditional::prune(
 
   // Check the max value for every combination of our keys.
   // If the max value is 0.0, we can prune the corresponding conditional.
-  auto pruner = [&](const Assignment<Key> &choices,
-                    const GaussianConditional::shared_ptr &conditional)
-      -> GaussianConditional::shared_ptr {
-    return (max->evaluate(choices) == 0.0) ? nullptr : conditional;
+  auto pruner =
+      [&](const Assignment<Key> &choices,
+          const GaussianFactorValuePair &pair) -> GaussianFactorValuePair {
+    if (max->evaluate(choices) == 0.0)
+      return {nullptr, std::numeric_limits<double>::infinity()};
+    else
+      return pair;
   };
 
-  auto pruned_conditionals = conditionals_.apply(pruner);
-  return std::make_shared<HybridGaussianConditional>(discreteKeys(),
-                                                     pruned_conditionals);
+  FactorValuePairs prunedConditionals = factors().apply(pruner);
+  return std::shared_ptr<HybridGaussianConditional>(
+      new HybridGaussianConditional(discreteKeys(), prunedConditionals));
 }
 
 /* *******************************************************************************/
 double HybridGaussianConditional::logProbability(
-    const HybridValues &values) const {
-  auto conditional = conditionals_(values.discrete());
-  return conditional->logProbability(values.continuous());
+    const HybridValues& values) const {
+  auto [factor, _] = factors()(values.discrete());
+  if (auto conditional = std::dynamic_pointer_cast<GaussianConditional>(factor))
+    return conditional->logProbability(values.continuous());
+  else
+    throw std::logic_error(
+        "A HybridGaussianConditional unexpectedly contained a non-conditional");
 }
 
 /* *******************************************************************************/
-double HybridGaussianConditional::evaluate(const HybridValues &values) const {
-  auto conditional = conditionals_(values.discrete());
-  return conditional->evaluate(values.continuous());
+double HybridGaussianConditional::evaluate(const HybridValues& values) const {
+  auto [factor, _] = factors()(values.discrete());
+  if (auto conditional = std::dynamic_pointer_cast<GaussianConditional>(factor))
+    return conditional->evaluate(values.continuous());
+  else
+    throw std::logic_error(
```

Comment:
I'm wondering if there is a better way to reduce the duplication of this and logProbability? Probably by passing in a functional?
E.g.
```cpp
double checkAndRun(const HybridValues& values, std::function<double(const GaussianConditional::shared_ptr& conditional, const HybridValues& values)>& functional) {
  auto [factor, _] = factors()(values.discrete());
  if (auto conditional = std::dynamic_pointer_cast<GaussianConditional>(factor))
    return functional(conditional, values);
  else
    throw std::logic_error(
        "A HybridGaussianConditional unexpectedly contained a non-conditional");
}
```

Reply:
For another day...

Reply:
I already worked on it so pushed the changes.